bftengine/src/bcstatetransfer/BCStateTran.cpp

```diff
@@ -917,12 +922,26 @@ void BCStateTran::startCollectingStateInternal() {
   startCollectingStateImpl();
 }
 
-void BCStateTran::startCollectingStateImpl() {
+void BCStateTran::setExternalCollectingState(bool newVal) {
+  if (newVal == externalCollectingState_) {
```

Comment:
This function is not thread-safe. The fact that externalCollectingState_ is atomic does not help here, because there are multiple operations on it: 'if' and then replace. There could be a race in between.

Reply:
you can read the content externalCollectingState_ and use compare_exchange_strong.
Not sure if this should be done only for the prints though.